collectors/cgroups.plugin/sys_fs_cgroup.c

```diff
@@ -859,7 +871,114 @@ struct discovery_thread {
+    return simple_pattern_matches(enabled_cgroup_paths, id);
+}
+
+static inline int matches_enabled_cgroup_names(char *name) {
+    return simple_pattern_matches(enabled_cgroup_names, name);
+}
+
+static inline int matches_enabled_cgroup_renames(char *id) {
+    return simple_pattern_matches(enabled_cgroup_renames, id);
+}
+
+static inline int matches_systemd_services_cgroups(char *id) {
+    return simple_pattern_matches(systemd_services_cgroups, id);
+}
+
+static inline int matches_search_cgroup_paths(const char *dir) {
+    return simple_pattern_matches(search_cgroup_paths, dir);
+}
+
+static inline int is_cgroup_systemd_service(struct cgroup *cg) {
+    return (cg->options & CGROUP_OPTIONS_SYSTEM_SLICE_SERVICE);
+}
+
+// ---------------------------------------------------------------------------------------------
+static int k8s_is_container(const char *id) {
+    // examples:
+    // https://github.com/netdata/netdata/blob/0fc101679dcd12f1cb8acdd07bb4c85d8e553e53/collectors/cgroups.plugin/cgroup-name.sh#L121-L147
+    const char *p = id;
+    const char *pp = NULL;
+    int i = 0;
+    size_t l = strlen("pod");
+    while ((p = strstr(p, "pod"))) {
+        i++;
+        p += l;
+        pp = p;
+    }
+    return !(i < 2 || !pp || !(pp = strchr(pp, '/')) || !pp++ || !*pp);
+}
+
+#define TASK_COMM_LEN 16
+
+static int k8s_get_container_first_proc_comm(const char *id, char *comm) {
+    if (!k8s_is_container(id)) {
+        return 1;
+    }
+
+    static procfile *ff = NULL;
+
+    char filename[FILENAME_MAX + 1];
+    snprintfz(filename, FILENAME_MAX, "%s/%s/cgroup.procs", cgroup_cpuacct_base, id);
+
+    ff = procfile_reopen(ff, filename, NULL, PROCFILE_FLAG_DEFAULT);
+    if (unlikely(!ff)) {
+        debug(D_CGROUP, "CGROUP: k8s_is_pause_container(): cannot open file '%s'.", filename);
+        return 1;
+    }
+
+    ff = procfile_readall(ff);
+    if (unlikely(!ff)) {
+        debug(D_CGROUP, "CGROUP: k8s_is_pause_container(): cannot read file '%s'.", filename);
+        return 1;
+    }
+
+    unsigned long lines = procfile_lines(ff);
+    if (likely(lines < 2)) {
+        return 1;
+    }
+
+    char *pid = procfile_lineword(ff, 0, 0);
+    if (!pid || !*pid) {
+        return 1;
+    }
+
+    snprintfz(filename, FILENAME_MAX, "%s/proc/%s/comm", netdata_configured_host_prefix, pid);
+
+    ff = procfile_reopen(ff, filename, NULL, PROCFILE_FLAG_DEFAULT);
+    if (unlikely(!ff)) {
+        debug(D_CGROUP, "CGROUP: k8s_is_pause_container(): cannot open file '%s'.", filename);
+        return 1;
+    }
+
+    ff = procfile_readall(ff);
+    if (unlikely(!ff)) {
+        debug(D_CGROUP, "CGROUP: k8s_is_pause_container(): cannot read file '%s'.", filename);
+        return 1;
+    }
+
+    lines = procfile_lines(ff);
+    if (unlikely(lines != 2)) {
+        return 1;
+    }
+
+    char *proc_comm = procfile_lineword(ff, 0, 0);
+    if (!proc_comm || !*proc_comm) {
+        return 1;
+    }
+
+    strncpy(comm, proc_comm, TASK_COMM_LEN);
```

Comment:
A suggestion to improve security, add as argument the size of argument `comm` for we never copy more data than expected.
I also suggest to add `comm[size] = 0` to be sure we won't have string without an end, because this function is assuming it will always receive a string fill with zeros.

Reply:
> comm[size] = 0

This is [done by the caller](https://github.com/ilyam8/netdata/blob/3345b14b6a54317a72e6e5a50aba307c6b454482/collectors/cgroups.plugin/sys_fs_cgroup.c#L2544).

> add as an argument the size of argument

Not sure about that. This is not a generic function, `/proc/PID/comm` is required to be 16

> Strings longer than TASK_COMM_LEN (16) characters (including the terminating null byte) are silently truncated

I can add an `if` check at the beginning of the function to check the length of the comm.

Reply:
Partially fixed by using `strncpyz`. Not sure we need to fix "for we never copy more data than expected".

Reply:
> This is [done by the caller](https://github.com/ilyam8/netdata/blob/3345b14b6a54317a72e6e5a50aba307c6b454482/collectors/cgroups.plugin/sys_fs_cgroup.c#L2544).

But, if the developer forgets? This was the motive I gave the suggestion to transfer the code.

> Not sure about that. This is not a generic function, /proc/PID/comm is required to be 16

PID is by default `32768` when distribution does not change the value when compiles the kernel, but on `Alma`, `Ubuntu`, and `Manjaro` we can have `4194304`, as you can see, we cannot assume the length. The cited values are inside `/proc/sys/kernel/pid_max`.

Reply:
An update when you wrote about `/proc/PID/comm` you were talking about the tasks not file name. :facepalm:  Sorry, you are right, it is always 16.